### DISCUSSION TITLE
Add glucose level as device class with unit converter

### DISCUSSION THREAD

**Discussion by epenet** [home-assistant/architecture#1146] [Entity Models] [answered]
We should support `mmol/L` and `mg/dL` for glucose level, and have the ability to move freely from one to the other.

Based on comment by @frenck in https://github.com/home-assistant/core/pull/126602#pullrequestreview-2330869859

This is a follow-up to https://github.com/home-assistant/core/pull/58921, https://github.com/home-assistant/core/pull/54898, and https://github.com/home-assistant/core/pull/41382 (all closed)

Other options considered:
- it should not be part of an option flow, as that would break statistics
- it should not be part of a config flow, as we now have reconfiguration and it would break statistics
- expose 2 sensors (also no really nice)

Currently only two components would use this: `dexcom`, `nightscout`

**Comment by frenck** [comment]
We've discussed this in the core meeting today, and agreed we are ok with a go-ahead on this one.

../Frenck

**Reply by epenet** [reply]
For implementation purpose, we need to have:
- a new device class
  - I propose that we add this as `DeviceClass.BLOOD_SUGAR_LEVEL`
- units added to global constants
  - I propose that we add a new `UnitOfVolumeConcentration` enumerator, as these units might in the future be used for things other than blood sugar level.
- a unit converter
  - Because of molecular weight, the converter needs to be specific to blood sugar level. I suggest `BloodSugarLevelConverter` for this purpose

**Comment by frenck** [comment]
Removed the accepted answer from the architectural discussion, as discussions started to raise after the accepted decision was communicated.

Need re-evaluation at a later moment.

**Comment by gagebenne** [comment]
What is the reasoning for not including these units under [concentration units](https://github.com/home-assistant/core/blob/d78fcd2a293f4855bc7b1ab1850d08d019302a84/homeassistant/const.py#L1287)? "Blood sugar level" is synonymous with "blood sugar concentration", and `mg/dL` (mass concentration) and `mmol/L` (molar concentration) are two concentrations units that are highly associated with glucose readings, but not exclusive to them.

As for freely converting between the two... A blood glucose integration user likely would never want to freely switch between `mg/dL` and `mmol/L`, only selecting the option on initial setup. This is the flow I would like the [`Dexcom`](https://github.com/home-assistant/core/tree/dev/homeassistant/components/dexcom) integration to move to. Currently it allows the switch after setup via the CONFIGURE button -- not good.

The conversion on the fly may not be worth any downsides that come with it. Broadly, blood glucose reading watchers learn their country's unit and that's that. In rare exceptions where a user needs both -- why not a template sensor?

**Reply by epenet** [reply]
> What is the reasoning for not including these units under [concentration units](https://github.com/home-assistant/core/blob/d78fcd2a293f4855bc7b1ab1850d08d019302a84/homeassistant/const.py#L1287)? "Blood sugar level" is synonymous with "blood sugar concentration", and `mg/dL` (mass concentration) and `mmol/L` (molar concentration) are two concentrations units that are highly associated with glucose readings, but not exclusive to them.

I think the (dis-)agreement is just on the naming of the constants:
- Specific enum? `UnitOfBloodSugarConcentration.XXX` / `UnitOfBloodSugarLevel.XXX` ...
- Regular constants? `CONCENTRATION_XXX`
- Split enum `UnitOfMassConcentration.XXX` + `UnitOfMolarConcentration.XXX`
- Generic enum for all volume concentrations (mixing mass/volume and parts or mmol/volume)? `UnitOfVolumeConcentration.XXX`
- Generic enum for all concentrations (mixing mass/volume and parts or mmol/volume and even ppm/ppb)? `UnitOfConcentration.XXX`

(see also https://github.com/home-assistant/architecture/discussions/1147)

**Reply by gagebenne** [reply]
Yeah, up to you folks whether to specify concentrations down to volume, mass, molar, etc. Just think when some other integration comes along that uses, say, `mg/dL`, it'd be strange to use a constant with blood glucose in the naming. Then do we create an additional generic `mg/dL` unit and have that confusion? It seems best not to pigeon-hole the unit and keep it generic.

**Reply by epenet** [reply]
That's exactly why I turned away from `UnitOfGlucoseLevel` when I started on the actual implementation.
(see https://github.com/home-assistant/core/pull/126820)

As it stands, I'm now leaning towards having two enums:
- `UnitOfDensity.XXX` https://en.wikipedia.org/wiki/Density
- `UnitOfMolarConcentration.XXX` https://en.wikipedia.org/wiki/Molar_concentration

**Reply by frenck** [reply]
We cannot offer generic calculations from mmol/L, because of the molecular weight that isn't a constant. That is why we agreed on a specific enum for glucose/blood sugar levels and did dismiss the idea of using it in a concentration enum.

../Frenck

**Reply by epenet** [reply]
> We cannot offer generic calculations from mmol/L, because of the molecular weight that isn't a constant. That is why we agreed on a specific enum for glucose/blood sugar levels and did dismiss the idea of using it in a concentration enum.
> 
> ../Frenck

But the UnitConverter do not need to be a match to a UnitOf*** constant.
We have UnitlessConverter as an example of this decoupling:
https://github.com/home-assistant/core/blob/d5ad35630f89faaae71ec748e871494470745d1e/homeassistant/util/unit_conversion.py#L558-L571

The UnitConverter is linked to a device class, so we can couple that easily:
`BloodSugarLevelConverter` <=> `DeviceClass.BLOOD_SUGAR_LEVEL`

**Reply by frenck** [reply]
It is easier to say (ignore exact name) `gluclose` device class, that can have any unit in `UnitOfGlucose` in which conversion between either in that enum can be switch and converted in between. The conclusion in the meeting yesterday was: We don't need to make it more difficult.

../Frenck

**Reply by gagebenne** [reply]
If a "converter" cannot be tied to a specific device class as @epenet was proposing... Another option is to not support conversion and decide the unit during configuration flow only. I think users need the choice, but nobody is flip-flopping day to day?

The units can stay generic, less work. They are really two entirely different ways to measure something, we just happen to know how to convert this specific instance.

Maybe `nightscout` code owner @marciogranzotto has different thoughts.

**Reply by frenck** [reply]
> Another option is to not support conversion and decide the unit during configuration flow only.

See OP for why we do not want to got down that path.

**Reply by epenet** [reply]
> It is easier to say (ignore exact name) `gluclose` device class, that can have any unit in `UnitOfGlucose` in which conversion between either in that enum can be switch and converted in between. The conclusion in the meeting yesterday was: We don't need to make it more difficult.
> 
> ../Frenck

If we have two enums, then the glucose device class could still use any unit from `{UnitOfMassConcentration + UnitOfMolarConcentration}`, including `μg/ft³`, `mg/m³` and `µg/m³` even if those are less used than `mmol/L` and `mg/dL`

**Reply by epenet** [reply]
Due to the lack of agreement on the concentration enumerators, I have proposed a fresh implementation in https://github.com/home-assistant/core/pull/129340

**Reply by emontnemery** [reply, marked as answer]
I think we should go for what @frenck summarizes here https://github.com/home-assistant/architecture/discussions/1146#discussioncomment-10774172, i.e.
- A blood glucose concentration sensor device class 
- A corresponding unit of blood glucose concentration unit enum
- Unit conversion framework between the units of glucose concentration

The exact names of the device class and unit enum can be nailed down in an implementation PR

If Home Assistant evolves in a direction where converting between molarity and mass concentrations for many substances is of importance, we should fix that then, instead of trying to solve that now before there's such a need.

**Reply by epenet** [reply]
I will do that, it that is what is desired, but a unit enum for "glucose concentration" goes against what was discussed on all other `UnitOf` enums, which were specifically linked to SI units (mass vs weight, etc.)

`UnitOf` enums were for developers (not user facing) whereas the device class was user facing.

**Reply by epenet** [reply]
Anyway - I have adjusted the suggested implementation PR to reflect this: https://github.com/home-assistant/core/pull/129340

**Reply by frenck** [reply]
We have discussed this one again. The reason we want to deviate from the normal procedures here (like the mass vs weight you pointed out) is because in the conversion between these units, the weight of the molecule matters. It is thus not universal; grouping this this way, makes it more focussed towards the purpose.

For now, let's go this route.
